perftest/perftest.go

```diff
@@ -0,0 +1,222 @@
+		if err != nil {
+			return "", err
+		}
+		voteOnProposal(propID)
+		time.Sleep(time.Second * 10)
+	}
+
+	// Move all markets out of auction
+	markets = getMarkets()
+	if len(markets) > 0 {
+		sendOrder(markets[0], 0, 10010, -100, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+		sendOrder(markets[0], 1, 9900, +100, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+		sendOrder(markets[0], 0, 10000, +5, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+		sendOrder(markets[0], 1, 10000, -5, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+	} else {
+		return "", fmt.Errorf("failed to get open market")
+	}
+	time.Sleep(time.Second * 5)
+
+	return markets[0], nil
+}
+
+func sendTradingLoad(marketID string, users, ops, runTimeSeconds int) error {
+	// Start load testing by sending off lots of orders at a given rate
+	userCount := users - 2
+	now := time.Now()
+	count := 0
+	delays := 0
+	ordersPerSecond := ops
+	opsScale := 1.0
+	if ordersPerSecond > 1 {
+		opsScale = float64(ordersPerSecond - 1)
+	}
+	// Work out how many orders we need for 10 minute run
+	numberOfTransactions := runTimeSeconds * ordersPerSecond
+	for i := 0; i < numberOfTransactions; i++ {
+		user := rand.Intn(userCount) + 2
+		choice := rand.Intn(100)
+		if choice < 2 {
+			// Perform a cancel all
+			sendCancelAll(user, marketID)
+		} else if choice < 7 {
+			// Perform a market order to generate some trades
+			if choice%2 == 1 {
+				sendOrder(marketID, user, 0, 3, "MARKET", proto.Order_TIME_IN_FORCE_IOC, 0)
+			} else {
+				sendOrder(marketID, user, 0, -3, "MARKET", proto.Order_TIME_IN_FORCE_IOC, 0)
+			}
+		} else {
+			// Insert a new order to fill up the book
+			priceOffset := rand.Int63n(12) - 6
+			if priceOffset > 0 {
+				// Send a sell
+				sendOrder(marketID, user, int64(10000+user), -1, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+			} else {
+				// Send a buy
+				sendOrder(marketID, user, int64(9999-user), 1, "LIMIT", proto.Order_TIME_IN_FORCE_GTC, 0)
+			}
+		}
+		count++
+
+		newNow := time.Now()
+		actualDiffSeconds := newNow.Sub(now).Seconds()
+		wantedDiffSeconds := float64(count) / opsScale
+
+		// See if we are sending quicker than we should
+		if actualDiffSeconds < wantedDiffSeconds {
+			delayMillis := (wantedDiffSeconds - actualDiffSeconds) * 1000
+			if delayMillis > 10 {
+				time.Sleep(time.Millisecond * time.Duration(delayMillis))
+				delays++
+			}
+
+			newNow = time.Now()
+			actualDiffSeconds = newNow.Sub(now).Seconds()
+		}
+
+		if actualDiffSeconds >= 1 {
+			fmt.Printf("\rSending load transactions...[%d/%d] %dcps  ", i, numberOfTransactions, count)
+			count = 0
+			delays = 0
+			now = newNow
+		}
+	}
+	fmt.Printf("\rSending load transactions...")
+	return nil
+}
+
+// Run is the main function of `perftest` package
+func Run(dataNodeAddr, walletURL, faucetURL, ganacheURL string, commandsPerSecond, runtimeSeconds, userCount int) error {
+	flag.Parse()
+
+	savedWalletURL = walletURL
+
+	fmt.Print("Connecting to data node...")
```

Comment:
maybe move this below the `len(dataNodeAddr) <= 0` check. We don't want to give users hope 😄